panels/models/attraction.py

```diff
@@ -569,28 +706,106 @@ def overlap(self, event):
 
 class AttractionSignup(MagModel):
     attraction_event_id = Column(UUID, ForeignKey('attraction_event.id'))
+    attraction_id = Column(UUID, ForeignKey('attraction.id'))
     attendee_id = Column(UUID, ForeignKey('attendee.id'))
 
-    signup_time = Column(UTCDateTime, server_default=utcnow())
-    checkin_time = Column(UTCDateTime, nullable=True)
+    signup_time = Column(UTCDateTime, default=lambda: datetime.now(pytz.UTC))
+    checkin_time = Column(
+        UTCDateTime, default=lambda: utcmin.datetime, index=True)
```

Comment:
Here is another interesting design choice made for the sake of performance. I made `checkin_time` not null, so it could be indexed. Indexes on nullable columns don't work when you are trying to select rows where the column is null.

Instead of using null to indicate "not checked in yet", I am using the date of "0001-01-01 0:00". This has really sped up some queries, but we can no longer use the Falseyness of `checkin_time` to indicate "not checked in".

Reply:
Indexing nullable columns still works if you're querying for a non-null value. For example, if you want to query for rows matching a given window, between start and end dates, the database will be able to use the index. However, if you're querying for rows with a null value, the database will have to do a table scan.

I am actually quite surprised by this behavior, and I only discovered it by running some queries against `EXPLAIN` in `psql`. I would have assumed that postgres would have some way of knowing which rows where _not_ contained in the index. Like, somehow the index would keep all the rows with null values in a set that it could return if you're querying for a null value.

Reply:
I definitely also assumed that null values on an indexed column would still be queryable without a full table scan - glad to know that's not the case!